### PR TITLE
[depends: #1063] Use nullptr to represent null pointers (targets master)

### PR DESCRIPTION
**Pull request by reuk**
This PR was created using clang-tidy's auto-fix mode. To run clang-tidy, I had to generate a compile database, which required a CMake build. I wonder whether it would be worth creating another PR to add CMake support to CBMC master, but *alongside* the existing makefile build. We could add a single configuration to Travis that builds using CMake instead, to ensure that both builds are kept up-to-date.

### CODE REVIEW

**Comment by tautschnig** [comment]
I wholeheartedly support this PR. Yet I am wondering how we best deal with the linter warnings - are you ok dealing with them or should I pitch in and add a commit that takes care of them?

**Comment by reuk** [comment]
I can fix them

**Comment by tautschnig** [comment]
Thanks!

**Comment by reuk** [comment]
Some of the linter complaints refer to bigint and miniz, as these modules were modified by the nullptr pass. As these are not strictly "our" code, I'm not sure how to proceed. Should the nullptr changes be rolled-back in these files, or should I leave them in place and fix the linter errors too?

**Comment by tautschnig** [comment]
> Should the nullptr changes be rolled-back in these files, or should I leave them in place and fix the linter errors too?

I'm slightly in favour of rolling them back, but also any other approach is fine with me.

**Comment by reuk** [comment]
Should the linter be adjusted to ignore "external" code like these modules?

File: src/analyses/invariant_set.h [hunk 1]
```diff
@@ -22,6 +22,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "interval_template.h"
 
+#define nullptr_exceptiont(str) str
```
Comment (tautschnig) [inline src/analyses/invariant_set.h]: I am wondering whether this is the best approach and right place to put?

File: src/analyses/reaching_definitions.h [hunk 2]
```diff
@@ -26,6 +26,9 @@ class is_threadedt;
 class dirtyt;
 class reaching_definitions_analysist;
 
+#define bad_cast_exceptiont(str) str
+#define nullptr_exceptiont(str) str
```
Comment (tautschnig) [inline src/analyses/reaching_definitions.h]: Now I'm convinced this isn't the right place.

File: src/analyses/invariant_set.h (hunk 1 above)
Comment (reuk) [inline src/analyses/invariant_set.h]: This is temporary, until #1063 gets merged. Of course this should be a custom exception type derived from `invariant_failedt`, and these common exception types should probably live in `util`. Should I add a "util/standard_exceptions.h" header and just put macros in it for now?

Comment (tautschnig) [inline src/analyses/invariant_set.h]: Would it be ok to wait for #1063 to get merged and then do it properly?

Comment (reuk) [inline src/analyses/invariant_set.h]: Yes, that's fine with me.

Comment (tautschnig) [inline src/analyses/invariant_set.h]: Changed the PR title in a desperate attempt to introduce dependency tracking into github...

**Comment by reuk** [comment]
This was not ready to be merged. It required fixes which depended on #1063.

**Comment by tautschnig** [comment]
@reuk Reverted after OOB comms with @kroening. Would you mind putting in an updated PR, now that #1063 is merged? Thanks!

**Comment by reuk** [comment]
Yes, I'll put together a new version once I have a chance.

**Pull request merged**